### ISSUE TITLE
Is there any plan to finish stargo and merge back to master?

### ISSUE THREAD

**Issue by yangzh** [google/starlark-go#220]
Hey, Thanks for this fantastic project, and I enjoyed/learned a lot!

I happened to come across the stargo branch, however, there is no activities since early this year, and I'm wondering what's the plan to move forward: it would be such a great tool to power Starlark packages with native Go code and I think it can have a lot of interesting usages.

Thanks.

**Comment by alandonovan** [comment]
I would like to merge it into the main tree, but it has not been a priority. 

The 'addressing' branch (containing changes to the compiler) cannot be committed as is, because it defines two new operators `&` (as in `&x[i]` or '&x.f`) and unary `*` (as in `*ptr` or `*bytes.Buffer`), and, convenient though these are syntactically for Go users, they don't fit well within the rest of the language. In particular, unary `*` already has a different meaning in a function call such as `f(*args)`. I propose to replace the operators by calls to built-in functions:
* `&x[i]` becomes `addressof(x[i])`;
* `*ptr` becomes `deref(ptr)`;
* `*bytes.Buffer` becomes `go.ptr_to(bytes.Buffer)`.
The `addressof` and `deref` functions require compiler support.

**Comment by mewmew** [comment]
Thanks for the update @adonovan. And thanks for working on the Starlark-go project in the first place!

For someone who is quite new to Starlark, could you help me clarify one thing?

In particular, I was wondering how the *stargo* branch (once finished and merged) differs in functionality from that provided by [starlight](https://github.com/starlight-go/starlight)?

From the [FAQ of starlight](https://github.com/starlight-go/starlight#why):

> **Why not just use starlark-go directly?**
>
> Well, it's actually quite difficult to get go data into a starlark-go script. Starlark as written is made more for configuration, so you mostly get data out of scripts, and mostly just basic values.

From what I can tell, it seems like the *stargo* branch has a similar purpose, to get Go values into Starlark scripts?

If this has been brought up before, sorry for the noise. I'm still reading on the Starlark spec and ecosystem up as I go.

Cheers,
Robin

**Edit:** while writing this post, I stumbled over the https://github.com/google/starlark-go/commit/3ffba8b164ff5bb67b8f3808c0a09595147764af commit, and it seems to explain the relationship between *stargo* and Starlight.

> This is essentially Nate Finch's Starlight-Go idea,
> but with an attempt to solve some of the aliasing problems
> in the Starlark compiler by introducing the concept of
> modes (addressable/var/lvalue versus nonaddressable/value/rvalue).

So, second question. Would Starlight become obsolete once *stargo* is in a stable enough shape to be merge into master?

**Comment by alandonovan** [comment]
On Fri, 12 Jul 2019 at 17:18, Robin Eklind <notifications@github.com> wrote:

> Thanks for the update @adonovan <https://github.com/adonovan>. And thanks
> for working on the Starlark-go project in the first place!
>
> For someone who is quite new to Starlark, could you help me clarify one
> thing?
>
> In particular, I was wondering how the *stargo* branch (once finished and
> merged) differs in functionality from that provided by starlight
> <https://github.com/starlight-go/starlight>?
>
> From the FAQ of starlight <https://github.com/starlight-go/starlight#why>:
>
> *Why not just use starlark-go directly?*
>
> Well, it's actually quite difficult to get go data into a starlark-go
> script. Starlark as written is made more for configuration, so you mostly
> get data out of scripts, and mostly just basic values.
>
> From what I can tell, it seems like the *stargo* branch has a similar
> purpose, to get Go values into Starlark scripts?
>
> If this has been brought up before, sorry for the noise. I'm still reading
> on the Starlark spec and ecosystem up as I go.
>
> Cheers,
> Robin
>
> *Edit:* while writing this post, I stumbled over the 3ffba8b
> <https://github.com/google/starlark-go/commit/3ffba8b164ff5bb67b8f3808c0a09595147764af>
> commit, and it seems to explain the relationship between *stargo* and
> Starlight.
>
> This is essentially Nate Finch's Starlight-Go idea,
> but with an attempt to solve some of the aliasing problems
> in the Starlark compiler by introducing the concept of
> modes (addressable/var/lvalue versus nonaddressable/value/rvalue).
>
>
Stargo is very similar to (and was inspired by) Starlight, but there are
some subtle and confusing semantic flaws in Starlight that cannot be fixed
without fundamental changes to the compiler, because Go (like C and C++)
allows expressions such as x, a[i], and x.f to have their address taken,
and they behave like addresses (not values) in that case. This requires
that the compiler generate different code for those expressions when their
address is taken (or when they appear on the LHS of an assignment). By
contrast, Starlark (like Python and Java) does not allow a variable to have
its address taken: to update a variable you must use a syntax like x=v,
a[i]=v, or x.f=v. This is explained in more detail in comments in the
'stargo' and 'addressing' git branches that you linked to.


> So, second question. Would Starlight become obsolete once *stargo* is in
> a stable enough shape to be merge into master?
>
Mostly, yes. Starlight does provide additional features such as convenient
helper functions for converting between Go and Starlark values. I don't
want to add them to the core because, convenient though they are, they are
logically external libraries, and they also make policy choices that are
not fully determined by the core.

**Comment by mewmew** [comment]
@alandonovan thanks for the impressive response time and thorough answer. I've started digging through the *stargo* branch. Turns out you post most of your rationale in the README. I should have looked a bit deeper, so thanks for taking the time to reply!

**Comment by mewmew** [comment]
As pointed out in the stargo README, there are several design principles that differ from those attributed to Starlark.

> - beware: stargo undermines many Starlark properties:
>  programs may be nondeterministic (e.g. due to map iteration)
>  programs may block (e.g. due to channels);
>  programs may panic (though we attempt to catch them);
>  values are not frozen, so data races are possible;
>  the lvalue/rvalue concept becomes important.

In particular, I'd be curious to learn, what are the considerations for *stargo* regarding **Hermetic execution**? I'm assuming that "anything goes" when using *stargo* as it gives mostly access to a large subset of the primitives and packages provided by Go. However, a quick test using the `net/http` module to access a `file://` scheme, results in an `unsupported protocol scheme "file"` error (as expected for Hermetic execution):

```
>>> resp, err = http.Get('file:///tmp/secret.txt')
>>> print(err)
Get file:///tmp/secret.txt: unsupported protocol scheme "file"
```

So, any specific thought regarding security considerations of embedding Starlark if given access to *stargo*?

**Comment by alandonovan** [comment]
> So, any specific thoughts regarding security considerations of embedding Starlark if given access to stargo?

Essentially it's up to the person exposing Go functions to Starlark to analyze the implications for determinism, security, resource usage, and so on.

The Stargo-wrapped HTTP package may report errors properly, but it is absolutely not hermetic: its result depends on the state of the Internet.

**Comment by mewmew** [comment]
> > So, any specific thoughts regarding security considerations of embedding Starlark if given access to stargo?
> 
> Essentially it's up to the person exposing Go functions to Starlark to analyze the implications for determinism, security, resource usage, and so on.
> 
> The Stargo-wrapped HTTP package may report errors properly, but it is absolutely not hermetic: its result depends on the state of the Internet.

Yeah, I realized after posting that it was a rather silly way to state the question. It does make sense that you provide the environment for Starlark to interact with and through, and given the exposure of that environment the security aspects of the specific embedded instance of Starlark may be evaluated.

This concept feels a lot like how processes are handed in [Genode](https://genode.org/), where a new process has access to almost nothing, but a handle to their parent through which they may ask to get access to and awareness of more of the outside world.

I started looking at the implementation of `stargo/cmd/stargo` and it looks very clean. Cool use of the `gen` tool. A more restricted environment would simply not provide access to e.g. `io/ioutil`.

**Comment by yangzh** [comment]
> I would like to merge it into the main tree, but it has not been a priority.
> 
> The 'addressing' branch (containing changes to the compiler) cannot be committed as is, because it defines two new operators `&` (as in `&x[i]` or `&x.f`) and unary `*` (as in `*ptr` or `*bytes.Buffer`), and, convenient though these are syntactically for Go users, they don't fit well within the rest of the language. In particular, unary `*` already has a different meaning in a function call such as `f(*args)`. I propose to replace the operators by calls to built-in functions:
> 
> * `&x[i]` becomes `addressof(x[i])`;
> * `*ptr` becomes `deref(ptr)`;
> * `*bytes.Buffer` becomes `go.ptr_to(bytes.Buffer)`.
>   The `addressof` and `deref` functions require compiler support.

Again, thanks for taking time to the thorough response. 
I also took a quick look at the "addressing" branch, it has no activities since early this year, approximately the same time as the 'stargo' branch. It's already 6 months.

I was a bit anxious as I want to understand how this project is managed / progressed under Google's Open Source program. Are you effectively the only active contributor (for authoring, reviewing and planning), or is there anybody (either inside Google or outside) helping out? It has been too many projects (from Google, or elsewhere) that went into limbo, due to insufficient fund / focus / priorities.

Thanks.

**Comment by alandonovan** [comment]
> I also took a quick look at the "addressing" branch, it has no activities since early this year, approximately the same time as the 'stargo' branch. It's already 6 months.

The addressing and stargo branches are intimately related; neither makes sense without the other.
They are essentially a prototype of a new feature for Starlark-in-Go. There has been some discussion about using Stargo to support dynamically loaded vet analyzers, but no definite plans yet, which is why this work is essentially on ice.


> I was a bit anxious as I want to understand how this project is managed / progressed under Google's Open Source program. Are you effectively the only active contributor (for authoring, reviewing and planning), or is there anybody (either inside Google or outside) helping out? It has been too many projects (from Google, or elsewhere) that went into limbo, due to insufficient fund / focus / priorities.

It's a good question. Starlark-in-Go is not an official Google product, but it is used by a number of teams within and outside Google, some of whom are regular contributors. I work on Google's build tools, where I develop (among other things) Starlark in general, including the Java implementation. The goals of the Starlark-in-Go project are quite conservative---essentially to maintain a simple and efficient implementation of the language spec, which changes very little, so its ongoing maintenance costs are low.  But development of new features (including Stargo, but also standard packages such as json, time, proto, etc) happens slowly.

**Comment by mewmew** [comment]
Hi @alandonovan,

I hope you've had a good start of 2020. Sweden didn't really have winter this year, but the arrival of spring always brings with joy and warmth.

As you mentioned, Starlark seem to be a quite stable language, so few updates have been needed to Starlark-in-Go. Still nice to see that there has been some ongoing development since last summer.

As someone who's still very excited for the integration of *Stargo* into Starlark-in-Go, I'd like to ask if there's been any update on the discussions the addressing feature in Starlark-in-Go which would enable *Stargo*?

Cheerful regards,
Robin

**Comment by alandonovan** [comment]
No, sorry, stargo has not moved forward since I last commented.

**Comment by FourSigma** [comment]
Any updates on the `stargo` branch? Any alternatives that that we could use? @alandonovan

**Comment by adonovan** [comment]
I had a number of discussions (some years ago now) with experts in both Starlark and Go, and our consensus was that Stargo would not be a good idea. Although there does appear to be a reasonable semantics for lvalues and rvalues, it is quite subtle; more importantly, the ability to do useful work, safely, in Stargo is quite limited by problems with error handling: many exceptional situations are reported by raising a Go panic, not a Starlark error, for example. 

So, fun though this was, I don't plan to spend more time on it. Sorry.
